File: src/main/java/com/dasbikash/news_server_spring_mvc_rest_datasource/SourceLineCounter.java
/*
 * Copyright 2019 das.bikash.dev@gmail.com. All rights reserved.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.dasbikash.news_server_spring_mvc_rest_datasource;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.RandomAccessFile;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public class SourceLineCounter {

    private static List<File> allFiles = new ArrayList<File>();
    private static COUNT_MODE countMode = COUNT_MODE.EXCLUDE_TEST_FILES;

    enum COUNT_MODE{
        ABSOLUTE_ALL,ALL,EXCLUDE_TEST_FILES
    }

    public static void main(String[] args) {
        File file;
        switch (countMode){
            case ALL:
            case ABSOLUTE_ALL:
                file = new File("src/");
                break;
            case EXCLUDE_TEST_FILES:
                file = new File("src/main/");
                break;
            default:
                return;
        }

        if (file.exists() && file.isDirectory() && file.canExecute()){
            processFile(file);
        }

        final HashMap<File,Integer> fileLineCountMap = new HashMap<>();

        allFiles.forEach(file1 -> {
            System.out.print("File: "+file1.getAbsolutePath());
            if (file1.getName().matches("\\..+")){
                System.out.println(" is hidden so quiting");
                return;
            }
            try {
                RandomAccessFile fileReader = new RandomAccessFile(file1,"r");
                int lineCount = 0;

                do {
                    String nextLine;
                    try {
                        nextLine = fileReader.readLine();
                    } catch (IOException e) {
                        e.printStackTrace();
                        break;
                    }
                    if (nextLine == null){
                        break;
                    }

                    if (countMode != COUNT_MODE.ABSOLUTE_ALL) {
                        if (nextLine.trim().isEmpty() ||
                                nextLine.contains("import") ||
                                nextLine.contains("package") ||
                                nextLine.matches("^//.+") ||
                                nextLine.matches("^.\\*.+")) {
                            continue;
                        }
                    }
                    lineCount++;
                }while (true);
                System.out.println(" Line count: "+lineCount);
                fileLineCountMap.put(file1,lineCount);
            } catch (FileNotFoundException e) {
                e.printStackTrace();
            }
        });

        int totalLineCount = 0;
        for (int lineCount :
                fileLineCountMap.values()) {
            totalLineCount += lineCount;
        }

        System.out.println("Total lines: "+totalLineCount);


    }


    private static void processFile(File file){
        String[] childFileNames= file.list();

        if (childFileNames!=null && childFileNames.length>0) {

            for (String childFileName : childFileNames) {
                File childFile = new File(file.getAbsolutePath()+"/"+childFileName);
                if (childFile.exists() && childFile.isDirectory() && childFile.canExecute()) {
                    processFile(childFile);
                } else {
                    allFiles.add(childFile);
                }
            }
        }
    }
}

File: src/main/kotlin/com/dasbikash/news_server_spring_mvc_rest_datasource/SourceLineCounter.kt
/*
 * Copyright 2019 das.bikash.dev@gmail.com. All rights reserved.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.dasbikash.news_server_spring_mvc_rest_datasource

import java.io.File
import java.io.FileNotFoundException
import java.io.IOException
import java.io.RandomAccessFile
import java.util.ArrayList
import java.util.HashMap

object SourceLineCounter {

    private val allFiles = ArrayList<File>()
    private val countMode = COUNT_MODE.EXCLUDE_TEST_FILES

    internal enum class COUNT_MODE {
        ABSOLUTE_ALL, ALL, EXCLUDE_TEST_FILES
    }

    @JvmStatic
    fun main(args: Array<String>) {
        val file: File
        when (countMode) {
            SourceLineCounter.COUNT_MODE.ALL, SourceLineCounter.COUNT_MODE.ABSOLUTE_ALL -> file = File("src/")
            SourceLineCounter.COUNT_MODE.EXCLUDE_TEST_FILES -> file = File("src/main/")
        }

        if (file.exists() && file.isDirectory && file.canExecute()) {
            processFile(file)
        }

        val fileLineCountMap = HashMap<File, Int>()

        allFiles.forEach { file1 ->
            print("File: " + file1.absolutePath)
            if (file1.name.matches("\\..+".toRegex())) {
                println(" is hidden so quiting")
                return//@allFiles.forEach
            }
            try {
                val fileReader = RandomAccessFile(file1, "r")
                var lineCount = 0

                do {
                    val nextLine: String?
                    try {
                        nextLine = fileReader.readLine()
                    } catch (e: IOException) {
                        e.printStackTrace()
                        break
                    }

                    if (nextLine == null) {
                        break
                    }

                    if (countMode != COUNT_MODE.ABSOLUTE_ALL) {
                        if (nextLine.trim { it <= ' ' }.isEmpty() ||
                                nextLine.contains("import") ||
                                nextLine.contains("package") ||
                                nextLine.matches("^//.+".toRegex()) ||
                                nextLine.matches("^.\\*.+".toRegex())) {
                            continue
                        }
                    }
                    lineCount++
                } while (true)
                println(" Line count: $lineCount")
                fileLineCountMap[file1] = lineCount
            } catch (e: FileNotFoundException) {
                e.printStackTrace()
            }
        }

        var totalLineCount = 0
        for (lineCount in fileLineCountMap.values) {
            totalLineCount += lineCount
        }

        println("Total lines: $totalLineCount")


    }


    private fun processFile(file: File) {
        val childFileNames = file.list()

        if (childFileNames != null && childFileNames.size > 0) {

            for (childFileName in childFileNames) {
                val childFile = File(file.absolutePath + "/" + childFileName)
                if (childFile.exists() && childFile.isDirectory && childFile.canExecute()) {
                    processFile(childFile)
                } else {
                    allFiles.add(childFile)
                }
            }
        }
    }
}
